core/src/main/java/hudson/model/Label.java

```diff
@@ -389,7 +389,7 @@ public int getTiedJobCount() {
                         for (Item i : parent.getItems()) {
                             if (i instanceof AbstractProject) {
                                 final AbstractProject project = (AbstractProject) i;
-                                if (this.equals(project.getAssignedLabel())) {
+                                if (this.name.equals(project.getAssignedLabelString())) {
```

Comment:
I think the correct way to do this is to introduce a matches method, e.g.

```
public boolean matches(String labelString) {
  return name.equals(labelString);
}

public final boolean equals(Object that) {
  if (this == that) return true;
  if (that == null || getClass() != that.getClass()) return false;
  return matches(((Label)that).name);
}
```

and then use `matches(project.getAssignedLabelString())` in place of `this.name.equals(project.getAssignedLabelString()` as it makes the contract clearer. It also leaves the door open for a subclass to use a different matching strategy.